pkg/state/state_secret.go

```diff
@@ -0,0 +1,112 @@
+package state
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+type secretSerializer struct {
+	secretNS   string
+	secretName string
+	secretKey  string
+	logger     log.Logger
+}
+
+func newSecretSerializer(logger log.Logger, ns string, name string, key string) stateSerializer {
+	return &secretSerializer{logger: logger, secretNS: ns, secretName: name, secretKey: key}
+}
+
+func (s *secretSerializer) Load() (State, error) {
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		return State{}, errors.Wrap(err, "get in cluster config")
+	}
+
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return State{}, errors.Wrap(err, "get kubernetes client")
+	}
+
+	if s.secretNS == "" {
```

Comment:
it looks like the secret serializer checks if all required parameters are set. Do we check that the get/put URLs are provided when state-from=url ?

Reply:
I kind of debated this.  There is no viper object here, so it's not really appropriate to hardcode command line arguments here.  And you still get a pretty meaningful error if the URL is an empty string.

Reply:
I think agree: in either case, doing validation inside the serializer implementation seems like the wrong place. Feels like if we add validation, it should probably be a bit higher up, where we pull stuff out of viper